pkg/reconciler/build/build.go

```diff
@@ -0,0 +1,136 @@
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/knative/pkg/controller"
+	"github.com/knative/pkg/logging"
+	"github.com/knative/pkg/logging/logkey"
+
+	clientset "github.com/knative/build/pkg/client/clientset/versioned"
+	buildscheme "github.com/knative/build/pkg/client/clientset/versioned/scheme"
+	informers "github.com/knative/build/pkg/client/informers/externalversions/build/v1alpha1"
+	listers "github.com/knative/build/pkg/client/listers/build/v1alpha1"
+)
+
+const controllerAgentName = "build-controller"
+
+// Reconciler is the controller.Reconciler implementation for Builds resources
+type Reconciler struct {
+	// kubeclientset is a standard kubernetes clientset
+	kubeclientset kubernetes.Interface
+	// buildclientset is a clientset for our own API group
+	buildclientset clientset.Interface
+
+	buildsLister listers.BuildLister
+
+	// Sugared logger is easier to use but is not as performant as the
+	// raw logger. In performance critical paths, call logger.Desugar()
+	// and use the returned raw logger instead. In addition to the
+	// performance benefits, raw logger also preserves type-safety at
+	// the expense of slightly greater verbosity.
+	Logger *zap.SugaredLogger
+}
+
+// Check that we implement the controller.Reconciler interface.
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func init() {
+	// Add build-controller types to the default Kubernetes Scheme so Events can be
+	// logged for build-controller types.
+	buildscheme.AddToScheme(scheme.Scheme)
+}
+
+// NewController returns a new build template controller
+func NewController(
+	logger *zap.SugaredLogger,
+	kubeclientset kubernetes.Interface,
+	buildclientset clientset.Interface,
+	buildInformer informers.BuildInformer,
+) *controller.Impl {
+
+	// Enrich the logs with controller name
+	logger = logger.Named(controllerAgentName).With(zap.String(logkey.ControllerType, controllerAgentName))
+
+	r := &Reconciler{
+		kubeclientset:  kubeclientset,
+		buildclientset: buildclientset,
+		buildsLister:   buildInformer.Lister(),
+		Logger:         logger,
+	}
+	impl := controller.NewImpl(r, logger, "Builds")
+
+	logger.Info("Setting up event handlers")
+	// Set up an event handler for when Build resources change
+	buildInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    impl.Enqueue,
+		UpdateFunc: controller.PassNew(impl.Enqueue),
+	})
+
+	// TODO(mattmoor): Set up a Pod informer, so that Pod updates
+	// trigger Build reconciliations.
+
+	return impl
+}
+
+// Reconcile implements controller.Reconciler
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logger.Errorf("invalid resource key: %s", key)
+		return nil
+	}
+
+	// Get the Build resource with this namespace/name
+	_, err = c.buildsLister.Builds(namespace).Get(name)
```

Comment:
nit:

```go
if _, err := c.buildsLister...; errors.IsNotFound(err) {
  // not found 
} else if err != nil {
  return err
}
```

Reply:
Done, although when we start acting on the other part we'll probably rehoist it.